File: src/handlers/page-view.ts
// Imports the Google Analytics Data API client library.
import { BetaAnalyticsDataClient, type protos } from "@google-analytics/data";
import { join } from "path";
import merge from "deepmerge";

const RESPONSE_MAX_AGE = 108000; // 30 hours in seconds

/**
 * TODO(developer): Uncomment this variable and replace with your
 *   Google Analytics 4 property ID before running the sample.
 */
const propertyId = "356543625";

const propertyIdMap = Object.freeze({
	"sahithyan.dev": "356543625",
	"kalvi.lk": "398313900",
});

function isPropertyAvailableForDomain(
	domain: string
): domain is keyof typeof propertyIdMap {
	return domain in propertyIdMap;
}

process.env.GOOGLE_APPLICATION_CREDENTIALS = join(
	process.cwd(),
	"./blog-pageviews-1713811580440-51ea0fe058d1.json"
);
console.log(
	"GOOGLE_APPLICATION_CREDENTIALS",
	process.env.GOOGLE_APPLICATION_CREDENTIALS
);

// Using a default constructor instructs the client to use the credentials
// specified in GOOGLE_APPLICATION_CREDENTIALS environment variable.
const analyticsDataClient = new BetaAnalyticsDataClient();

const headers = {
	"Content-Type": "application/json",
	"Cache-Control": `public, max-age=${RESPONSE_MAX_AGE}`,
};

export async function GET(
	request: Request,
	responseInitOptions: ResponseInit = {}
) {
	const url = new URL(request.url);
	const pagePath = url.searchParams.get("path");

	let targetDomain = url.searchParams.get("domain");
	if (!targetDomain) {
		request.headers.forEach((value, key) => {
			console.log("header", key, value);
		});
		const originHeader = request.headers.get("Origin");
		if (originHeader) {
			targetDomain = originHeader.split(":")[0].split("://")[1];
		}
	}

	console.log("getting page views for", targetDomain, "on", pagePath);
	if (!targetDomain || !isPropertyAvailableForDomain(targetDomain)) {
		return new Response(
			"Unsupported domain",
			merge<ResponseInit>(
				{
					status: 400,
				},
				responseInitOptions
			)
		);
	}

	const runReportOptions: protos.google.analytics.data.v1beta.IRunReportRequest =
		{
			property: `properties/${propertyIdMap[targetDomain]}`,
			dimensions: [{ name: "pagePath" }],
			metrics: [{ name: "screenPageViews" }],
			dateRanges: [{ startDate: "2022-03-05", endDate: "yesterday" }],
		};
	if (pagePath) {
		runReportOptions.dimensionFilter = {
			filter: {
				fieldName: "pagePath",
				stringFilter: { matchType: "EXACT", value: pagePath },
			},
		};
	}

	const responseArr = await analyticsDataClient
		.runReport(runReportOptions)
		.catch(console.error);
	const response = responseArr === undefined ? undefined : responseArr[0];
	const rows = response?.rows;
	if (!rows) {
		console.log(response);
		return new Response(
			"Not available now",
			merge<ResponseInit>(
				{
					status: 500,
				},
				responseInitOptions
			)
		);
	}

	const map: Record<string, string> = {};
	const _responseOptions = merge<ResponseInit>(
		{
			status: 200,
			headers,
		},
		responseInitOptions
	);
	console.log("merged response options", _responseOptions);

	for (const row of rows) {
		if (!row.dimensionValues || !row.metricValues) {
			continue;
		}
		const key = row.dimensionValues[0].value;
		const value = row.metricValues[0].value;
		if (!pagePath && typeof key === "string" && typeof value === "string") {
			map[key] = value;
		} else if (pagePath === key) {
			return new Response(
				JSON.stringify({
					views: value || "0",
					dateRetrievedOn: new Date().toISOString(),
				}),
				_responseOptions
			);
		}
	}

	return new Response(
		JSON.stringify({
			pages: map,
			dataRetrievedOn: new Date().toISOString(),
		}),
		_responseOptions
	);
}

File: internal/page-view.go
package internal

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	analyticsdata "google.golang.org/api/analyticsdata/v1beta"
)

const RESPONSE_MAX_AGE = 108600

var propertyIdMap = map[string]string{
	"sahithyan.dev":    "356543625",
	"kalvi.lk":         "398313900",
	"s1.sahithyan.dev": "449008877",
}

func GetPageView(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query()

	targetDomain := query.Get("domain")
	if targetDomain == "" {
		originHeader := r.Header.Get("Origin")
		fmt.Println("origin", originHeader)

		if originHeader == "" {
			w.Header().Set("status", "400")
			io.WriteString(w, "Couldn't find target domain")
			return
		}

		targetDomain = strings.Split(strings.Replace(originHeader, "https://", "", 1), ":")[0]
	}
	pagePath := query.Get("path")

	targetPropertyId, isDomainPresent := propertyIdMap[targetDomain]
	if !isDomainPresent {
		w.Header().Set("status", "400")
		io.WriteString(w, fmt.Sprintf("Unsupported domain: %s", targetDomain))
		return
	}

	os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", "./blog-pageviews-1713811580440-51ea0fe058d1.json")
	ctx := context.Background()
	analyticsdataService, err := analyticsdata.NewService(ctx)

	if err != nil {
		fmt.Printf("Couldn't start analyticsdata service: %s\n", err)
		w.Header().Set("status", "500")
		io.WriteString(w, "Internal error occured")
		return
	}

	runReportRequest := &analyticsdata.RunReportRequest{
		Dimensions: []*analyticsdata.Dimension{
			{
				Name: "pagePath",
			},
		},
		DateRanges: []*analyticsdata.DateRange{
			{
				StartDate: "2022-03-05",
				EndDate:   "today",
			},
		},
		Metrics: []*analyticsdata.Metric{
			{
				Name: "screenPageViews",
			},
		},
		OrderBys: []*analyticsdata.OrderBy{
			{
				Desc: true,
				Metric: &analyticsdata.MetricOrderBy{
					MetricName: "screenPageViews",
				},
			},
		},
	}
	if pagePath != "" {
		runReportRequest.DimensionFilter = &analyticsdata.FilterExpression{
			Filter: &analyticsdata.Filter{
				FieldName: "pagePath",
				StringFilter: &analyticsdata.StringFilter{
					MatchType: "EXACT",
					Value:     pagePath,
				},
			},
		}

	}

	result, err := analyticsdataService.Properties.RunReport(fmt.Sprintf("properties/%s", targetPropertyId), runReportRequest).Do()

	if err != nil {
		fmt.Printf("Couldn't run report on analyticsdata: %s\n", err)
		w.Header().Set("status", "500")
		io.WriteString(w, "Internal error occured")
		return
	}

	pageViewMap := map[string]string{}

	for i := 0; i < int(result.RowCount); i++ {
		row := result.Rows[i]
		key := row.DimensionValues[0].Value
		value := row.MetricValues[0].Value

		if pagePath == key {
			w.Header().Set("Content-Type", "application/json")
			w.Header().Set("Cache-Control", fmt.Sprintf("public, max-age=%d", RESPONSE_MAX_AGE))
			io.WriteString(w, fmt.Sprintf(`{"views": "%s", "dateRetrievedOn": "%s" }`, value, time.Now().UTC().Format(time.RFC3339)))
			return
		}

		pageViewMap[key] = value
	}
	output, err := json.Marshal(pageViewMap)
	if err != nil {
		fmt.Printf("Error occured while marshalling json: %s\n", err)
		w.Header().Set("status", "500")
		io.WriteString(w, "Internal error occured")
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", fmt.Sprintf("public, max-age=%d", RESPONSE_MAX_AGE))
	io.WriteString(w, string(output))
}
